cpp/src/arrow/extension/opaque_test.cc

```diff
@@ -0,0 +1,189 @@
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "arrow/extension/opaque.h"
+#include "arrow/extension_type.h"
+#include "arrow/io/memory.h"
+#include "arrow/ipc/reader.h"
+#include "arrow/ipc/writer.h"
+#include "arrow/record_batch.h"
+#include "arrow/testing/extension_type.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/type_fwd.h"
+#include "arrow/util/checked_cast.h"
+
+namespace arrow {
+
+TEST(OpaqueType, Basics) {
+  auto type = internal::checked_pointer_cast<extension::OpaqueType>(
+      extension::opaque(null(), "type", "vendor"));
+  auto type2 = internal::checked_pointer_cast<extension::OpaqueType>(
+      extension::opaque(null(), "type2", "vendor"));
+  ASSERT_EQ("arrow.opaque", type->extension_name());
+  ASSERT_EQ(*type, *type);
+  ASSERT_NE(*arrow::null(), *type);
+  ASSERT_NE(*type, *type2);
+  ASSERT_EQ(*arrow::null(), *type->storage_type());
+  ASSERT_THAT(type->Serialize(), ::testing::Not(::testing::IsEmpty()));
+  ASSERT_EQ(R"({"type_name":"type","vendor_name":"vendor"})", type->Serialize());
+  ASSERT_EQ("type", type->type_name());
+  ASSERT_EQ("vendor", type->vendor_name());
+  ASSERT_EQ(
+      "extension<arrow.opaque[storage_type=null, type_name=type, vendor_name=vendor]>",
+      type->ToString(false));
+}
+
+TEST(OpaqueType, Equals) {
+  auto type = internal::checked_pointer_cast<extension::OpaqueType>(
+      extension::opaque(null(), "type", "vendor"));
+  auto type2 = internal::checked_pointer_cast<extension::OpaqueType>(
+      extension::opaque(null(), "type2", "vendor"));
+  auto type3 = internal::checked_pointer_cast<extension::OpaqueType>(
+      extension::opaque(null(), "type", "vendor2"));
+  auto type4 = internal::checked_pointer_cast<extension::OpaqueType>(
+      extension::opaque(int64(), "type", "vendor"));
+  auto type5 = internal::checked_pointer_cast<extension::OpaqueType>(
+      extension::opaque(null(), "type", "vendor"));
+
+  ASSERT_EQ(*type, *type);
+  ASSERT_EQ(*type2, *type2);
+  ASSERT_EQ(*type3, *type3);
+  ASSERT_EQ(*type4, *type4);
+  ASSERT_EQ(*type5, *type5);
+
+  ASSERT_EQ(*type, *type5);
+
+  ASSERT_NE(*type, *type2);
+  ASSERT_NE(*type, *type3);
+  ASSERT_NE(*type, *type4);
+
+  ASSERT_NE(*type2, *type);
+  ASSERT_NE(*type2, *type3);
+  ASSERT_NE(*type2, *type4);
+
+  ASSERT_NE(*type3, *type);
+  ASSERT_NE(*type3, *type2);
+  ASSERT_NE(*type3, *type4);
+
+  ASSERT_NE(*type4, *type);
+  ASSERT_NE(*type4, *type2);
+  ASSERT_NE(*type4, *type3);
```

Comment:
Perhaps also compare with an extension type that is not Opaque?

Reply:
Updated.